### ISSUE TITLE
Server side mod support?

### ISSUE THREAD

**Issue by Shadowblitz16** [OpenStarbound/OpenStarbound#4]
Not sure if steam sends the mod's of a server to a client but it would be nice if this wasn't just steam only.

**Comment by Bottinator22** [comment]
The game does not download mods for you. You have to match mod lists yourself.

**Comment by Shadowblitz16** [comment]
Sorry to rephrase...
I'm not sure if steam automatically sends mods from servers but it would be nice if it could be implemented in starbound.

I think tiy said at some point he wanted to do server side mods but it was never implemented.

**Comment by SweptWasTaken** [comment]
> Sorry to rephrase... I'm not sure if steam automatically sends mods from servers but it would be nice if it could be implemented in starbound.

Yeah so then someone could package a malicious file as a mod and then upload it to your PC the second you join a server.

**Comment by Shadowblitz16** [comment]
> > Sorry to rephrase... I'm not sure if steam automatically sends mods from servers but it would be nice if it could be implemented in starbound.
> 
> Yeah so then someone could package a malicious file as a mod and then upload it to your PC the second you join a server.

That's what sandboxing is for, and lua is perfect for that.
Devs should not be able to access anything outside of starbounds api.

Also another alternative is server side gameplay but that would mean alot of network data.

**Comment by Kawa-oneechan** [comment]
Having the server send its own mods, _especially_ with some notion of which mods came from which server so they don't clog up your single-player session or introduce mismatches with other servers,, would solve a _lot_ of client-server incompatibility issues I think.

Problem of course is if you want to, for example, play as a specific custom species but the server you're trying to connect to doesn't have that species mod.

**Comment by Shadowblitz16** [comment]
> Having the server send its own mods, _especially_ with some notion of which mods came from which server so they don't clog up your single-player session or introduce mismatches with other servers,, would solve a _lot_ of client-server incompatibility issues I think.
> 
> Problem of course is if you want to, for example, play as a specific custom species but the server you're trying to connect to doesn't have that species mod.

The server would be the authority. So no you wouldn't be able to play a species that the server doesn't have.
But that's not a problem because you could just ask them to add it or host your own server.

Also the server owner may not want some species especially if the server is meant for children.

**Comment by Kawa-oneechan** [comment]
And of course the opposite: say I don't like Frackin', find a server that seems interesting, but they have Frackin' for the sake of those who *do* like it. Sorry kid, but you're gonna have to contend with it.

Importantly, there should not suddenly now be a Frackin' installation on my local single player.

**Comment by Shadowblitz16** [comment]
> And of course the opposite: say I don't like Frackin', find a server that seems interesting, but they have Frackin' for the sake of those who _do_ like it. Sorry kid, but you're gonna have to contend with it.
> 
> Importantly, there should not suddenly now be a Frackin' installation on my local single player.

I really don't see your point minetest and king arthurs gold have such a thing

Make mods selectable in game.
Then you can choose what mods you want enabled by default.

-OR- 

There could be the case where mods are stored per universe and we can have multiple universes. and the universe data gets sent over when needed.

maybe a folder called universe could be our local copy and a folder called servers could store remote copies.

**Comment by Kawa-oneechan** [comment]
I feel like you might've missed my point.

If I join a server that has Frackin' installed, and other people who don't share my negative opinion actually use it, I will have to see Frackin' stuff. And even if I choose to endure that out of pure necessity, that server's copy of Frackin' better not be installed on my client when I leave.

You can't make server-side mods selectable, either -- if I unchecked the Frackin' installed on that server, and visited another player who didn't and uses it, what do you suppose will happen then? The exact thing server-side mods were supposed to fix, I'd say.
Or let's say it's a server full of species mods and you don't like cats. I'm there, you're there, and I'm playing a felin. We meet up, purely by chance, or you run into a felin settlement, or wandering trader. The same thing happens again. Server-side mods _cannot_ be made selectable by the clients.

**Comment by Shadowblitz16** [comment]
@Kawa-oneechan what about a checkbox while joining a server for uninstalling the mod's once you leave?
Then the next time you join a server the checkbox remembers your last option?

Also ye I meant mods not universes.

**Comment by Kawa-oneechan** [comment]
Methinks putting the mods in a separate, temporary folder would be better.

The only problem then becomes that the game has already loaded all of the _local_ assets, and joining a server would require not just downloading but then _reloading_ a completely different asset list. And characters are created before joining a game no matter if it's local or not.

That seems to me to be the bigger problem with server-provided mods, and probably a good reason we never got it from CF.

Seems like kind of a big deal breaker and I'm surprised and ashamed I didn't think of it back in January.

**Comment by Shadowblitz16** [comment]
> Methinks putting the mods in a separate, temporary folder would be better.
> 
> The only problem then becomes that the game has already loaded all of the _local_ assets, and joining a server would require not just downloading but then _reloading_ a completely different asset list. And characters are created before joining a game no matter if it's local or not.
> 
> That seems to me to be the bigger problem with server-provided mods, and probably a good reason we never got it from CF.
> 
> Seems like kind of a big deal breaker and I'm surprised and ashamed I didn't think of it back in January.

That's why they would be stored per server if the uninstall mods after leave is unchecked.

There really isn't a way to both have access to the mods and not have them on your computer.
You could stream mod content over the network when needed but that would make network requirements sky rocket

**Comment by Kawa-oneechan** [comment]
You seem to have missed the point.

Now:
1. You start the game.
2. It loads all assets, including the Workshop and `mods` folder. Takes a fair bit of time.
3. You make a character, if you haven't already.
4. You join a server, be it a local game or actually some other system.

With server side mods:
1. You start the game.
2. It loads all assets.
3. You make a character, maybe.
4. You join someone else's server.
5. The mod lists don't match so the server's are downloaded.
6. The game has to reload a different set of assets, from scratch. You sit and wait, again.
7. You have to make a new character because your fav species or something in your inventory isn't available anymore.
8. You have to go back to step six all over again whenever you change servers, including local play.

Does that help?

**Comment by Shadowblitz16** [comment]
@Kawa-oneechan yes I know this can be fixed in a couple of ways.

1. Multiplayer character could be server bound.
2. Character creation can be moved to after joining a server and future characters could store mod data and be filtered out based on server mods.

**Comment by Kawa-oneechan** [comment]
Now I _know_ for a fact you missed the point. Try to focus on points 2, 5, 6, and 8. Especially 2, 6, and 8.

**Comment by Shadowblitz16** [comment]
> Now I _know_ for a fact you missed the point. Try to focus on points 2, 5, 6, and 8. Especially 2, 6, and 8.

Yes this is how server mod's work, Would you rather spend 30 seconds waiting for starbound to download the mods or spend 3 hours looking up what mod's the server has, what version, what patches, downloading them and moving them to the mod's folder?

**Comment by Kawa-oneechan** [comment]
Again, focus on points 2, 6, and 8. They're basically the one same point repeated, that should've made comprehending them easier.

I wasn't talking about *download* times, which may indeed be only once per server. I was talking about *reloading* times. The fatness of your internet connection doesn't matter when the game takes a long time to *actually process* a complete collection of assets. That's bound to your CPU and hard drive and such.

All the assets, reloaded from ground zero. The base game, plus whatever mods apply to the currently selected server (or your local /`mods` folder and Workshop). From scratch, every time you switch servers, including going back to local, and when you open the game in the first place.

The big secret here is that if you want to fix this, you'd have to completely rewrite a rather significant chunk of the game's asset loading logic. I for one can't imagine any of the lead OpenSB folks doing something *that* daring. I can't imagine *me* trying, and I *certainly* can't imagine you doing it.

Unless you'd like to prove me wrong.

**Comment by HappyGoFishing** [comment]
what if you there could be a mod repository for OpenStarbound, like factorio has one built in?

**Comment by Kawa-oneechan** [comment]
_That's a whole other thing entirely._

Also there's like four or five mod repositories I could name in the wild already and probably more. Might I suggest this hypothetical in-game browser present an aggregate view instead of being a completely new one? Besides, if it was a new one, someone would have to _host_ the mods.

(CF website, Steam Workshop, Nexus, probably GameBanana, ~~LoversLab~~... yeah, I don't think we need yet another.)

**Comment by S0LiD-ST4T3** [comment]
if this hasn't been said before, now it has been.

I have an idea for the issue of having a mod installed locally - you can simply place assets into a TEMPORARY directory (/starbound/tmp/(server_address)/mods) when the player syncs mods, and only load that for the session on a particular server. once the player disconnects, the game reloads and ignores that "/tmp/" directory, rendering that theoretical copy of Frackin' to be Inert.

...or just stream data from the server as needed.

I believe the Source(tm) engine has the ability to cache assets locally while connected to a multiplayer server. Team Fortress 2 has been seen to acquire maps, models and Textures from servers and storing them locally in a cache folder before hotloading the newly acquired assets and joining a round.

...just one way to do it. heh.

**Comment by Kawa-oneechan** [comment]
You're not wrong. But again, your named example was _made_ to support such a thing while OpenSB would have to have a significant part rewritten to support hot patching like that.

Hotter than whatever `/reload` does, that is.

**Comment by Shadowblitz16** [comment]
Mods could also be listed before you join a server, so if you don't want fracken universe on your system you could just avoid that server.

Or again instead of client side mods if you don't want your mods to be shared, open Starbound could move/support mods to be server side and then the clients would never receive those mods only interact with server messages.

There are lots of ways to do something like this.
Alot of people didn't like the server side mods idea due to the amount of work it would take and due to it breaking vanilla Starbound compatibility and having two systems would be more work to maintain.

**Comment by Shadowblitz16** [comment]
I closed this because it would be too difficult to stream assets or state to clients.

Also automaticly downloading mods would work for making multiplayer easier.

And cheating can be disallowed by enforcing disallow asset mismatch.

The original idea for this was so people could have private mods that werent downloaded to peoples computers.
